### PR TITLE
Add from_base64url Presto Function

### PR DESCRIPTION
**Pull request by ashokku2022**

### CODE REVIEW

**Comment by ashokku2022** [comment]
resolves https://github.com/facebookincubator/velox/issues/4336

**Comment by ashokku2022** [comment]
@aditi-pandit have squashed my commits in one. Please take a look when you get time. Thanks!

**Comment by majetideepak** [comment]
Fuzzer test output from the description

```
I20230324 13:26:54.754741 8440631 ExpressionVerifier.cpp:68] All results match.
I20230324 13:26:54.754796 8440631 ExpressionFuzzer.cpp:1172] ==============================> Done with iteration 9013
I20230324 13:26:54.754842 8440631 ExpressionFuzzer.cpp:972] ==============================> Top 1 by number of rows processed
I20230324 13:26:54.754859 8440631 ExpressionFuzzer.cpp:974] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230324 13:26:54.754873 8440631 ExpressionFuzzer.cpp:978] from_base64url 9014 100.00% 314120
I20230324 13:26:54.754889 8440631 ExpressionFuzzer.cpp:984] ==============================> Bottom 1 by number of rows processed
I20230324 13:26:54.754902 8440631 ExpressionFuzzer.cpp:986] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230324 13:26:54.754916 8440631 ExpressionFuzzer.cpp:991] from_base64url 9014 100.00% 314120
I20230324 13:26:54.754931 8440631 ExpressionFuzzer.cpp:1004] ==============================> All stats sorted by number of times the function was chosen
I20230324 13:26:54.754945 8440631 ExpressionFuzzer.cpp:1006] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230324 13:26:54.754959 8440631 ExpressionFuzzer.cpp:1010] from_base64url 9014 100.00% 314120
```

File: velox/common/encode/Base64.cpp [hunk 1]
```diff
@@ -433,6 +433,15 @@ std::string Base64::encode_url(const folly::IOBuf* data) {
   return encodeImpl(IOBufWrapper(data), kBase64UrlCharset, false);
 }
 
+void Base64::decodeUrl(
```
Comment (majetideepak) [inline velox/common/encode/Base64.cpp:436]: `decode_url` similar to others?

File: velox/common/encode/Base64.h [hunk 2]
```diff
@@ -101,8 +102,19 @@ class Base64 {
   static size_t
   decode(const char* src, size_t src_len, char* dst, size_t dst_len);
 
- private:
+  static void
+  decode_url(const char* src, size_t src_len, char* dst, size_t dst_len);
```
Comment (majetideepak) [inline velox/common/encode/Base64.h:106]: where is this used?

Comment (ashokku2022) [reply velox/common/encode/Base64.h]: Seems like I missed it while cherry-pick. I have cleaned it up now, squashed and pushed the changes. Ran the fuzzer test again, output is as follow:

```
I20230327 10:00:03.238423 8996625 ExpressionFuzzer.cpp:974] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230327 10:00:03.238437 8996625 ExpressionFuzzer.cpp:978] from_base64url 9859 100.00% 342482
I20230327 10:00:03.238456 8996625 ExpressionFuzzer.cpp:984] ==============================> Bottom 1 by number of rows processed
I20230327 10:00:03.238471 8996625 ExpressionFuzzer.cpp:986] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230327 10:00:03.238484 8996625 ExpressionFuzzer.cpp:991] from_base64url 9859 100.00% 342482
I20230327 10:00:03.238499 8996625 ExpressionFuzzer.cpp:1004] ==============================> All stats sorted by number of times the function was chosen
I20230327 10:00:03.238513 8996625 ExpressionFuzzer.cpp:1006] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I20230327 10:00:03.238528 8996625 ExpressionFuzzer.cpp:1010] from_base64url 9859 100.00% 342482
```

File: velox/functions/lib/string/StringImpl.h [hunk 3]
```diff
@@ -427,6 +427,23 @@ FOLLY_ALWAYS_INLINE void toBase64Url(
...
+template <typename TOutString, typename TInString>
+FOLLY_ALWAYS_INLINE void fromBase64Url(
+    TOutString& output,
+    const TInString& input) {
+  try {
+    auto inputSize = input.size();
+    const bool hasPad =
+        (*(input.end()) == encoding::Base64::kBase64Pad) ? true : false;
+    output.resize(encoding::Base64::calculateDecodedSize(
+        input.data(), inputSize, hasPad));
```
Comment (aditi-pandit) [inline velox/functions/lib/string/StringImpl.h:439]: Nit : Maybe abstract a variable for input.data() also.

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 4]
```diff
@@ -1748,6 +1748,31 @@ TEST_F(StringFunctionsTest, fromBase64) {
   EXPECT_THROW(fromBase64("YQ==="), VeloxUserError);
 }
 
+TEST_F(StringFunctionsTest, fromBase64Url) {
+  const auto fromHex = [&](std::optional<std::string> value) {
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1752]: Please move this function definition closer to its usage.

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp [hunk 5]
```diff
@@ -1748,6 +1748,31 @@ TEST_F(StringFunctionsTest, fromBase64) {
...
+  EXPECT_EQ("a", fromBase64Url("YQ=="));
+  EXPECT_EQ("a", fromBase64Url("YQ"));
+  EXPECT_EQ("abc", fromBase64Url("YWJj"));
+  EXPECT_EQ("hello world", fromBase64Url("aGVsbG8gd29ybGQ="));
+  EXPECT_EQ(
+      "Hello World from Velox!",
+      fromBase64Url("SGVsbG8gV29ybGQgZnJvbSBWZWxveCE="));
+
+  EXPECT_EQ(fromHex("FF4FBF50"), fromBase64Url("_0-_UA=="));
+  EXPECT_THROW(fromBase64Url("YQ="), VeloxUserError);
```
Comment (aditi-pandit) [inline velox/functions/prestosql/tests/StringFunctionsTest.cpp:1770]: Its hard to follow why these are errors (since it seems like very similar values map correctly) ? Please can you add an explanation

File: velox/functions/lib/string/StringImpl.h (hunk 3 above)
Comment (ashokku2022) [reply velox/functions/lib/string/StringImpl.h]: done

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp (hunk 4 above)
Comment (ashokku2022) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1752]: It is close to the function definition, not sure what you meant. The function and test is within one place. Please clarify if it is not acceptable.

File: velox/functions/prestosql/tests/StringFunctionsTest.cpp (hunk 5 above)
Comment (ashokku2022) [reply velox/functions/prestosql/tests/StringFunctionsTest.cpp:1772]: Comment added to clarify, thanks!

File: velox/common/encode/Base64.h [hunk 6]
```diff
@@ -101,8 +102,16 @@ class Base64 {
...
 
- private:
+  static void decodeUrl(
+      const char* src,
+      size_t src_len,
+      char* dst,
+      size_t dst_len,
+      bool pad);
+
   constexpr static char kBase64Pad = '=';
```
Comment (aditi-pandit) [inline velox/common/encode/Base64.h:112]: Does this variable need to be public ? Seems like you moved the private keyword below it.

**Review by aditi-pandit — APPROVED** [review]

File: velox/common/encode/Base64.h (hunk 6 above)
Comment (ashokku2022) [reply velox/common/encode/Base64.h:112]: Yes, as I am using this variable `kBase64Pad` in `stringImpl.h` in my function to access it rather than using the '='. 
```
const bool hasPad =
        (*(input.end()) == encoding::Base64::kBase64Pad) ? true : false;
```

File: velox/common/encode/Base64.cpp [hunk 7]
```diff
@@ -433,6 +433,15 @@ std::string Base64::encode_url(const folly::IOBuf* data) {
...
+    const char* src,
+    size_t src_len,
+    char* dst,
+    size_t dst_len,
+    bool hasPad) {
+  decodeImpl(src, src_len, dst, dst_len, kBase64UrlReverseIndexTable, hasPad);
```
Comment (majetideepak) [inline velox/common/encode/Base64.cpp:442]: The existing `decode_url` on line 445 estimates and resizes the output. Why is this implementation different?

Comment (ashokku2022) [reply velox/common/encode/Base64.cpp:442]: the line 454 function signature has padding as false hardcoded, however, if you look at the rfc 4648 sec-5, it says, padding is optional, meaning - there could be padding or may not be padding present in the encoded input string, thus have to come up with new decodeUrl signature.

`[RFC 4648 §5](https://datatracker.ietf.org/doc/html/rfc4648#section-5): base64url (URL- and filename-safe standard)[[a]](https://en.wikipedia.org/wiki/Base64#cite_note-common-7)`

Comment (majetideepak) [reply velox/common/encode/Base64.cpp:442]: Does this mean the existing implementation has a bug? We can probably fix that in a follow-up PR by reusing this logic.

**Review by majetideepak — APPROVED** [review]

File: velox/common/encode/Base64.cpp (hunk 7 above)
Comment (ashokku2022) [reply velox/common/encode/Base64.cpp:442]: Based on the RFC, seems like there is a bug in the existing implementation as optional padding is not taken into account. Sure, will open a new PR to get that fixed.

**Comment by ashokku2022** [comment]
@mbasmanova Could you please review and merge when you get time, thank you!

**Review by mbasmanova — COMMENTED** [review]
@Yuhta @laithsakka Jimmy, Laith, would you help with this PR?

File: velox/common/encode/Base64.cpp [hunk 8]
```diff
@@ -433,7 +433,20 @@ std::string Base64::encodeUrl(const folly::IOBuf* data) {
   return encodeImpl(IOBufWrapper(data), kBase64UrlCharset, false);
 }
 
+<<<<<<< HEAD
```
Comment (Yuhta) [inline velox/common/encode/Base64.cpp:436]: There are still some conflicts

Comment (ashokku2022) [reply velox/common/encode/Base64.cpp]: @Yuhta my apology, IDE caused some issues, showed the conflict is gone but internally it stays so had to do it on the terminal. I have squashed and pushed, should be good now, thanks!

**Comment by Yuhta** [comment]
@ashokku2022 Can you rebase to newest master to fix the build signals?

**Comment by ashokku2022** [comment]
@Yuhta done!

**Comment by ashokku2022** [comment]
@Yuhta is it ready to merge?

File: velox/docs/functions/presto/binary.rst [hunk 9]
```diff
@@ -62,6 +62,10 @@ Binary Functions
 
     Encodes ``binary`` into a base64 string representation using the `URL safe alphabet <https://www.rfc-editor.org/rfc/rfc4648#section-5>`_.
 
+.. function:: from_base64url(string) -> varbinary
```
Comment (Yuhta) [inline velox/docs/functions/presto/binary.rst:65]: Can we move this entry to preserve the alphabetical order?

Comment (ashokku2022) [reply velox/docs/functions/presto/binary.rst]: @Yuhta Absolutely! done with changes, thanks for all your valuable feedback, highly appreciate it!

**Comment by ashokku2022** [comment]
@Yuhta please let me know if there is anything stopping this PR from merge.

**Comment by aditi-pandit** [comment]
> @Yuhta please let me know if there is anything stopping this PR from merge.

@ashokku2022 : Jimmy has merged the PR into Meta repositories already. The PR will get pushed by their automation once it completes tests etc.
